Source/Layout/ASCornerLayoutSpec.mm

```diff
@@ -0,0 +1,169 @@
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+
+#import <AsyncDisplayKit/ASCornerLayoutSpec.h>
+#import <AsyncDisplayKit/ASLayout.h>
+#import <AsyncDisplayKit/ASLayoutSpec+Subclasses.h>
+#import <AsyncDisplayKit/ASDisplayNode.h>
+
+CGPoint as_calculatedCornerOriginIn(CGRect baseFrame, CGSize cornerSize, ASCornerLayoutLocation cornerLocation, CGPoint offset) {
+  
+  CGPoint cornerOrigin = CGPointZero;
+  CGPoint baseOrigin = baseFrame.origin;
+  CGSize baseSize = baseFrame.size;
+  
+  switch (cornerLocation) {
+    case ASCornerLayoutLocationTopLeft:
+      cornerOrigin.x = baseOrigin.x - cornerSize.width / 2;
+      cornerOrigin.y = baseOrigin.y - cornerSize.height / 2;
+      break;
+    case ASCornerLayoutLocationTopRight:
+      cornerOrigin.x = baseOrigin.x + baseSize.width - cornerSize.width / 2;
+      cornerOrigin.y = baseOrigin.y - cornerSize.height / 2;
+      break;
+    case ASCornerLayoutLocationBottomLeft:
+      cornerOrigin.x = baseOrigin.x - cornerSize.width / 2;
+      cornerOrigin.y = baseOrigin.y + baseSize.height - cornerSize.height / 2;
+      break;
+    case ASCornerLayoutLocationBottomRight:
+      cornerOrigin.x = baseOrigin.x + baseSize.width - cornerSize.width / 2;
+      cornerOrigin.y = baseOrigin.y + baseSize.height - cornerSize.height / 2;
+      break;
+  }
+  
+  cornerOrigin.x += offset.x;
+  cornerOrigin.y += offset.y;
+  
+  return cornerOrigin;
+}
+
+static NSUInteger const kBaseChildIndex = 0;
+static NSUInteger const kCornerChildIndex = 1;
+
+@interface ASCornerLayoutSpec()
+@end
+
+@implementation ASCornerLayoutSpec
+
+- (instancetype)initWithChild:(id <ASLayoutElement>)child corner:(id <ASLayoutElement>)corner location:(ASCornerLayoutLocation)location
+{
+  self = [super init];
+  if (self) {
+    self.child = child;
+    self.corner = corner;
+    self.cornerLocation = location;
+  }
+  return self;
+}
+
++ (instancetype)cornerLayoutSpecWithChild:(id <ASLayoutElement>)child corner:(id <ASLayoutElement>)corner location:(ASCornerLayoutLocation)location
+{
+  return [[self alloc] initWithChild:child corner:corner location:location];
+}
+
+#pragma mark - Children
+
+- (void)setChild:(id<ASLayoutElement>)child
+{
+  ASDisplayNodeAssertNotNil(child, @"Child shouldn't be nil.");
+  [super setChild:child atIndex:kBaseChildIndex];
+}
+
+- (id<ASLayoutElement>)child
+{
+  return [super childAtIndex:kBaseChildIndex];
+}
+
+- (void)setCorner:(id<ASLayoutElement>)corner
+{
+  ASDisplayNodeAssertNotNil(corner, @"Corner element cannot be nil.");
+  [super setChild:corner atIndex:kCornerChildIndex];
+}
+
+- (id<ASLayoutElement>)corner
+{
+  return [super childAtIndex:kCornerChildIndex];
+}
+
+#pragma mark - Calculation
+
+- (ASLayout *)calculateLayoutThatFits:(ASSizeRange)constrainedSize
```

Comment:
This whole layout algorithm is a beauty!

Reply:
Great to hear that 😁